blogposts/inside-rust/01-announcement.md

```diff
@@ -0,0 +1,162 @@
+will be introduced specifically to support unwinding.
+
+## Three specific proposals
+
+The project group has narrowed the design space down to three specific
+proposals. Two of these introduce the new `"C unwind"` API, and one does not.
+
+Each proposal specifies the behavior of each type of unwind (Rust `panic!`,
+foreign (e.g. C++), and forced (e.g. `pthread` cancel)) when it encounters an
+ABI boundary under either the `panic=unwind` or `panic=abort` compile-mode.
+
+### 2 APIs, minimal specification
+
+* `"C"` ABI boundary, `panic=<any>`
+  * `panic!`: program aborts
+  * forced unwind, no destructors: unwind behaves normally
+  * all other foreign unwinding: undefined behavior
+* `"C unwind"` ABI boundary
+  * With `panic=unwind`: all types of unwinding behave normally
+  * With `panic=abort`: all types of unwinding abort the program
+
+### 2 APIs, forced unwinding always permitted
+
+This is the same as the previous design, except that when compiled with
+`panic=abort`, forced exceptions would *not* be intercepted; that is, the
+program would not abort.
+
+### 1 API
+
+* `panic=unwind`: unwind behaves normally
+* `panic=abort`:
+  * `panic!`: program aborts
+  * forced unwind, no destructors: unwind behaves normally
+  * all other foreign unwinding: undefined behavior
+
+## Comparison table for the proposed designs
+
+In this table, "UB" stands for "undefined behavior". Some instances of UB can
+be detected at runtime, but the code to do so would impose an undesirable
+code-size penalty; the group recommends generating such code only for debug
+builds. These cases are marked "UB (debug: abort)"
+
+Note that forced unwinding always causes undefined behavior when it traverses
+stack frames with destructors.
+
+|                               | “2 APIs” (Options 1 and 1c), “C”, “panic=unwind” | “2 APIs” (Options 1 and 1c), “C”, “panic=abort” | “2 APIs” (Options 1 and 1c), “C unwind”, “panic=unwind” | “2 APIs, always permit forced” (Option 1), “C unwind”, “panic=abort” | “2 APIs, minimal spec” (Option 1c), “C unwind”, “panic=abort” | “1 API” (Option 3), “panic=unwind” | “1 API” (Option 3), “panic=abort” |
+| ----------------------------- | ------------------------------------------------ | ----------------------------------------------- | ------------------------------------------------------- | -------------------------------------------------------------------- | ------------------------------------------------------------- | ---------------------------------- | --------------------------------- |
+| panic!                        | abort                                            | abort                                           | unwind                                                  | abort                                                                | abort                                                         | unwind                             | aborts                            |
```

Comment:
First box should be unwind, not abort.

Reply:
I don't believe so; part of the reason to introduce a second ABI is to _prevent_ unwinds (even `panic!`s) from crossing `extern "C"` boundaries.

Reply:
Oh! I got very confused by your use of `panic!`. You're not talking about the behavior of `panic!` itself, you're talking about *an exception generated by `panic!`*. This should be made clearer in the wording.

Also you should use N/A in the table for the panic=abort cases, since `panic!` does not generate an exception in those cases.

Reply:
That's an excellent point. Yes, the rows are different types of unwinding, so the `panic!` row is indeed for the unwinding operation, not the behavior of `panic!()`.

For `panic=abort`, it seems immaterial to me from a user perspective that the abort occurs before the `panic!()` macro actually starts unwinding the stack; in either case, the behavior is well defined and the ultimate outcome is that the program aborts. But you're right that the difference between "abort and function boundary" and "abort without unwinding" is observable in practice depending on what destructors are executed before the unwind operation reaches (or would reach) an `extern "C"` API boundary.

So perhaps the entries for that column should say something like "aborts without any unwinding"?

Reply:
I've reworded; it should be clearer.